controlpanel/frontend/forms.py

```diff
@@ -176,20 +193,64 @@ def __init__(self, *args, **kwargs):
 
 class CreateParameterForm(forms.Form):
     key = forms.CharField(
-        validators=[RegexValidator(r'[a-zA-Z0-9_]{1,50}')]
+        max_length=50,
+        validators=[
+            RegexValidator(
+                r'[a-zA-Z0-9_]{1,50}',
+                message="Must contain only alphanumeric characters and underscores",
+            ),
+        ],
     )
     role_name = forms.CharField(
-        validators=[RegexValidator(r'[a-zA-Z0-9_-]{1,60}')]
+        max_length=60,
+        validators=[
+            RegexValidator(
+                r'[a-zA-Z0-9_-]{1,60}',
+                message="Must contain only alphanumeric characters, underscores and hyphens",
+            ),
+        ],
     )
     value = forms.CharField(widget=forms.PasswordInput)
     app_type = forms.ChoiceField(choices=APP_TYPE_CHOICES)
 
 
 class CreateIAMManagedPolicyForm(forms.Form):
     name = forms.CharField(
+        # TODO restrict allowed characters in group policy name
         validators=[RegexValidator(POLICY_NAME_REGEX)]
     )
 
 
 class AddUserToIAMManagedPolicyForm(forms.Form):
     user_id = forms.CharField(max_length=128)
+
+
+class AppCustomersField(forms.Field):
+
+    def __init__(self, *, delimiters=APP_CUSTOMERS_DELIMITERS, strip=True, **kwargs):
+        self.delimiters = delimiters
+        self.strip = strip
+        super().__init__(**kwargs)
+
+    def to_python(self, value):
+        emails = self.delimiters.split(value)
+        if self.strip:
+            emails = [email.strip() for email in emails]
+        return emails
+
+    def clean(self, value):
+        value = self.to_python(value)
+        for email in value:
+            try:
+                validate_email(email)
+            except ValidationError:
```

Comment:
what's wrong with the message `validate_email` gives? I don't think you need to change it

Reply:
`validate_email` just says `Enter a valid email address.` - but the user may have entered a long list of email addresses, and it's helpful to know (the first one) which is invalid.